blog/2025-04-01-global-replication/index.mdx

```diff
@@ -0,0 +1,283 @@
+
+Let’s see what happens when a user uploads a file to a bucket:
+
+![](./upload-tigris-2.webp)
+
+The user uploads the picture of Rick Astely and its corresponding metadata.
+These two are separately handled. The picture is put into block storage (and
+maybe the SSD cache), but the metadata is stored directly in FoundationDB. Then
+the metadata is queued for replication.
+
+![](./upload-image-and-metadata-in-tigris.webp)
+
+A backend service handles our replication model. When it sees a new record in
+the replication queue, it eagerly pushes out the metadata to every other region.
+
+![](./pushing-out-metadata.webp)
+
+The really cool part about how this works under the hood is that the database is
+itself the message queue. Time as an ordered phenomenon\*. FoundationDB is an
+ordered datastore. The replication queue entries use the time that the object
+was created in its key name.
+
+:::note
+
+\*Okay yes there’s issues like time dilation when you’re away from a large
+source of mass like the earth (this is noticeable in the atomic clocks that run
+GPS in low earth orbit), and if you’re on a spaceship that’s near the speed of
+light. However, I’m talking about time in a vacuum with a nearby source of great
+mass, perfectly spherical cows, and whatnot, so it’s really not an issue for
+this example.
+
+:::
+
+This database-as-a-queue is based on how
+[iCloud's global replication works](https://www.foundationdb.org/files/QuiCK.pdf).
+It gives us a couple key advantages compared to using something like postgres
+and kafka:
+
+1. Data can be stored and queued for replication in the same transaction,
+   meaning that we don’t have to coordinate transactional successes and failures
+   between two systems
+2. Tigris is already an expert in running FoundationDB, so we can take advantage
+   of that experience and share it with our message queue, making this a lot
+   less complicated in practice.
+
+This isn’t a free lunch, there’s one sharp edge that you may run into: that
+replication takes a nonzero amount of time. It usually takes single digit
+seconds at most, which is more than sufficient for most applications. We’re
+working on ways to do better though\!
+
+## The secret fourth tier of caching
+
+Remember how I said that Tigris has three tiers of caching: block storage, SSD
+cache, and inline FoundationDB rows? There’s actually a secret fourth tier of
+caching: other Tigris regions. This is the key to how Tigris makes your data
+truly global.
+
+Let’s say you upload the pic of Rick to San Jose and someone requests it from
+Chicago. First, the data is put into San Jose’s block storage layer and the
+metadata is queued for replication.
+
+![](./secret-hack-1.webp)
+
+There’s a dirty trick going on in the metadata, let’s double click on it:
+
+```
+# Metadata:
+
+Name: rickastley.jpg
+Size: 63,178 bytes
+Cache: forever
+Regions: SJC
+```
+
+Every bit of metadata contains a reference to block storage. Block storage runs
```

Comment:
I'd cut down the line that says about different provider i.e.,
`Every bit of metadata contains a reference to block storage. The cool part is
that any Tigris region can pull from the block storage service in every other
region. Then it stores it inside the cache layer like normal.`

Reply:
Yeah, I would cut out the different provider part.

Say something like this - block storage runs in a different provider but in the same Tigris regions.